### ISSUE TITLE
retrieve_num keeps dropping

### ISSUE THREAD

**Issue by syr-cn** [RUCAIBox/R1-Searcher#21]
Hi authors, thank you for you great work.

I followed your the steps of your README to train a Qwen 2.5 with stage2 data and nq corpus, but the model fails to learn to actively retrieve documents during inference. 

Below is the log of some of my experiments:

![Image](https://github.com/user-attachments/assets/2ae745e2-f30c-48e0-9b6c-05d43c4bc891)

As you can see, the `retrieve_num` of the models keep dropping, and the response length curve is the same.

The detailed CLI args I used are listed below:
```
actor_init_on_gpu = False
actor_learning_rate = 2e-06
actor_num_gpus_per_node = 4
actor_num_nodes = 1
adam_betas = (0.9, 0.95)
adam_offload = False
advantage_estimator = reinforce_baseline
apply_chat_template = False
apply_query_filter = False
apply_select_response_by_prm = False
apply_select_response_longer_pos = False
apply_uncompleted_filter = False
aux_loss_coef = 0
bf16 = True
ckpt_path = Reinforce-Qwen2.5-7B-Zero
colocate_actor_ref = True
colocate_critic_reward = False
critic_learning_rate = 9e-06
critic_num_gpus_per_node = 8
critic_num_nodes = 1
critic_pretrain = None
disable_fast_tokenizer = False
enable_ema = False
enable_prefix_caching = False
enforce_eager = False
eps_clip = 0.2
eval_steps = -1
flash_attn = True
freezing_actor_steps = -1
gamma = 1
generate_max_len = 29000
grad_accum_dtype = None
gradient_checkpointing = True
gradient_checkpointing_use_reentrant = False
group_method = None
init_kl_coef = 0.0
input_key = question
input_template = None
kl_target = None
l2 = 0.0
lambd = 0.95
load_checkpoint = True
load_in_4bit = False
local_log_file = log/Reinforce-Qwen2.5-7B-Zero/log.jsonl
local_rank = -1
logging_steps = 1
lora_alpha = 16
lora_dropout = 0
lora_rank = 0
lr_warmup_ratio = 0.0
max_ckpt_mem = 100000000.0
max_ckpt_num = 3
max_epochs = 1
max_len = None
max_norm = 1.0
max_samples = 100000
micro_rollout_batch_size = 1
micro_train_batch_size = 1
n_samples_per_prompt = 16
normalize_reward = False
num_episodes = 10000
overlap_comm = True
packing_samples = True
perf = False
pretrain = model/Qwen2.5-7B
pretrain_data = None
pretrain_data_probs = 1.0
pretrain_split = train
prompt_data = data/training_set/stage_2.jsonl
prompt_data_probs = 1.0
prompt_max_len = 1024
prompt_split = train
ptx_coef = 0.05
random_temperature = False
ref_num_gpus_per_node = 4
ref_num_nodes = 1
ref_reward_offload = False
remote_rm_url = http://localhost:1278/get_reward
reward_clip_range = (-10, 10)
reward_num_gpus_per_node = 8
reward_num_nodes = 1
reward_pretrain = None
rollout_batch_size = 64
save_path = results/ckpts/Reinforce-Qwen2.5-7B-Zero
save_steps = 20
save_value_network = False
seed = 42
target_modules = all-linear
temperature = 1.0
top_p = 1.0
train_batch_size = 256
use_kl_estimator_k3 = False
use_kl_loss = False
use_length_reward_in_efficiency = False
use_tensorboard = None
use_wandb = 8c63841d0875e4fde65a42fb47b52e6a18b8a1ed
value_clip = 0.2
value_head_prefix = score
vllm_num_engines = 1
vllm_sync_backend = nccl
vllm_tensor_parallel_size = 1
wandb_group = None
wandb_org = None
wandb_project = R1-Searcher
wandb_run_name = Reinforce-Qwen2.5-7B-Zero
zero_stage = 2
zpg = 1
```

**Comment by SsmallSong** [comment]
It seems fine from just the training parameters. You can take a look at the data obtained during the rollout process in training to better pinpoint the issue.

**Comment by qishenghu** [comment]
Hi @syr-cn , I also encounter such issue, have u resolve this?

**Comment by syr-cn** [comment]
> Hi [@syr-cn](https://github.com/syr-cn) , I also encounter such issue, have u resolve this?

Yes. I found there are some bugs in the reward function in `R1-Searcher/train/reward_server_qwen_zero.py`. The format score only ask the counts of special to be the same, but it forgets to ask the counts to be non-zero.

Formally,
> if count_1 == count_2 == count_3 == count_4 == count_5 == count_6 == count_7:
>     pass
> else:
>     format_punishment=True

Should be:
> if count_1 == count_2 == count_3 == count_4 == count_5 == count_6 == count_7 >= 1:
>     pass
> else:
>     format_punishment=True

**Comment by SsmallSong** [comment]
> > Hi [@syr-cn](https://github.com/syr-cn) , I also encounter such issue, have u resolve this?
> > [@qishenghu](https://github.com/qishenghu)
> 
> Yes. I found there are some bugs in the reward function in `R1-Searcher/train/reward_server_qwen_zero.py`. The format score only ask the counts of special to be the same, but it forgets to ask the counts to be non-zero.
> 
> Formally,
> 
> > if count_1 == count_2 == count_3 == count_4 == count_5 == count_6 == count_7:
> > pass
> > else:
> > format_punishment=True
> 
> Should be:
> 
> > if count_1 == count_2 == count_3 == count_4 == count_5 == count_6 == count_7 >= 1:
> > pass
> > else:
> > format_punishment=True

Hello, will adopting this reward method lead to worse performance? Actually, for qwen-base, it's not necessary to explicitly specify a non-zero number of retrievals; the version I'm using is the publicly released one.

**Comment by syr-cn** [comment]
> > > Hi [@syr-cn](https://github.com/syr-cn) , I also encounter such issue, have u resolve this?
> > > [@qishenghu](https://github.com/qishenghu)
> > 
> > 
> > Yes. I found there are some bugs in the reward function in `R1-Searcher/train/reward_server_qwen_zero.py`. The format score only ask the counts of special to be the same, but it forgets to ask the counts to be non-zero.
> > Formally,
> > > if count_1 == count_2 == count_3 == count_4 == count_5 == count_6 == count_7:
> > > pass
> > > else:
> > > format_punishment=True
> > 
> > 
> > Should be:
> > > if count_1 == count_2 == count_3 == count_4 == count_5 == count_6 == count_7 >= 1:
> > > pass
> > > else:
> > > format_punishment=True
> 
> Hello, will adopting this reward method lead to worse performance? Actually, for qwen-base, it's not necessary to explicitly specify a non-zero number of retrievals; the version I'm using is the publicly released one.

In fact, I do witness a model collapse using the `counts>=1` setting.

![Image](https://github.com/user-attachments/assets/7542a49a-5d1f-4476-8070-d052f0f49191)

However, as I mentioned at the start of this issue, the training will fail at the very beginning if I don't set `counts>=1`. Thus, I can not provide a comparison between these two implementations.

**Comment by SsmallSong** [comment]
I think it might be an issue with the data curriculum. I've encountered a similar problem recently. 

I tried randomly shuffling the data, and under the same code, there were cases where the number of retrievals dropped to zero right at the beginning of training. 

This might be because if difficult questions are learned too early, the model might fail to answer them correctly during retrieval, which could lead to penalties due to formatting errors.

**Comment by SsmallSong** [comment]
You can add my WeChat, and we can communicate.

**Comment by WangYuzzu** [comment]
I encountered the same problem as you, which manifested as follows:
1. After a few steps of training, the model suddenly stopped using <search> and retrieve_num became 0
2. The length of the response suddenly became shorter

I am checking whether it is caused by the retrieval library

我和您遇到了相同的问题，表现为：
1. 模型在训练几步之后突然不再使用<search>，retrieve_num变为0
2. response的长度骤然变短

我正在检查是不是检索库导致的问题

<img width="1847" alt="Image" src="https://github.com/user-attachments/assets/ef8a6d1a-19c8-493a-88db-e017f690519a" />
